### ISSUE TITLE
Construction is too slow for 32M vertices

### ISSUE THREAD

**Issue by kvark** [hugoledoux/startin#24]
I'm trying to feed 32M vertices, and it's taking way too long. Any way to speed it up?
Would it make sense, for example, to split it into tiles and run multiple instances of `Triangulation` on different threads?

**Comment by hugoledoux** [comment]
Define "way too long"? 2min? 5h?

And different threads is not really possible with the code as is: we use one large triangulation and splitting it into sub-parts is not trivial...

What is your input? LAS? GeoTIFF?

you can always have a look at my other project https://github.com/hugoledoux/sst that allows you to create the TIN of billions of points, but it's not in Python...

**Comment by kvark** [comment]
"Too long" means I ran out of patience to wait for it. I was looking to get the results online, i.e. get it within seconds for the whole map, and then iteratively update it in real time. My input is raw data from a game - it's a height map of 2k by 16k size.

**Comment by hugoledoux** [comment]
then use `dt.insert(pts, insertionstrategy="BBox")` it should be several order of magnitude faster (not <1s though).

https://startinpy.readthedocs.io/latest/api.html#startinpy.DT.insert

see that example, a GeoTIFF is a height map:
https://startinpy.readthedocs.io/latest/examples.html#reading-a-geotiff-file-with-rasterio

**Comment by zdila** [comment]
Hello.

In my case I am inserting ~120M points from LAZ files in a loop using `insert_one_pt` and this way it is indeed very slow - I've interrupted the process after an hour. Then I tried to batch points by 1M and use `insert` with `InsertionStrategy::BBox` but it did not help.

@hugoledoux would you please propose me a better strategy? Maybe split my points to a grid and process every grid in a separate thread?

My source code: https://github.com/FreemapSlovakia/dm-maker

I will appreciate any help!

**Comment by hugoledoux** [comment]
With >100M points you're entering territories for which startinpy is perhaps not your best bet anymore... Batches could help, but if your file has a high spatial coherence (in short: if two points are close in the file are they also close in xy-coordinates?) then it will not change anything.

The insertion with strategy::bbox will not help one bit either.

Does your file has good spatial coherence? If yes then that's it, you can't do more I'm afraid. If not, then you could sort the points according to a Hilbert curve, OR use https://startinpy.readthedocs.io/latest/api.html#startinpy.DT.jump_and_walk (`dt.jump_and_walk = True`) which will help with the most expensive step of the algo (point location).

But perhaps you should try instead: 
- https://pypi.org/project/triangle/ 
- in C++, the CGAL lib: https://doc.cgal.org/latest/Triangulation_2/index.html#Section_2D_Triangulations_Delaunay
- in Rust: https://docs.rs/delaunator/latest/delaunator/

Hope this helps.

**Comment by hugoledoux** [comment]
oh sorry I see this was for the Rust code and not the Python bindings...

Well everything I said above is still valid actually. But if you're okay with C++ you can use Triangle directly perhaps, easier: https://www.cs.cmu.edu/~quake/triangle.html

**Comment by zdila** [comment]
Thank you @hugoledoux for your advice 👍. My points are not much sorted as I read in parallel from various flight lines. So I will try your recommendations.

**Comment by hugoledoux** [comment]
What you do is indeed deadly for a triangulator...

The jump-and-walk samples a certain number of points for each insertion, I have put a number there based on a known paper but you can perhaps play with this (more points means more Euclidean tested, but you could find a closer one...). Just play with that line: https://github.com/hugoledoux/startin/blob/master/src/lib.rs#L1224
